components/blitz/src/omero/cmd/fs/UsedFilesRequestI.java

```diff
@@ -0,0 +1,369 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+package omero.cmd.fs;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import loci.formats.FormatException;
+import loci.formats.IFormatReader;
+
+import org.hibernate.Query;
+import org.hibernate.Session;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Sets;
+
+import ome.io.nio.PixelsService;
+import ome.model.core.Image;
+import ome.model.core.Pixels;
+import ome.system.Login;
+import ome.util.SqlAction;
+import omero.cmd.ERR;
+import omero.cmd.Helper;
+import omero.cmd.IRequest;
+import omero.cmd.Response;
+import omero.cmd.UsedFilesRequest;
+import omero.cmd.HandleI.Cancel;
+import omero.cmd.UsedFilesResponse;
+import omero.cmd.UsedFilesResponsePreFs;
+import omero.constants.annotation.file.ORIGINALMETADATA;
+import omero.constants.namespaces.NSCOMPANIONFILE;
+
+/**
+ * Lists the IDs of the original files associated with an image.
+ * @author m.t.b.carroll@dundee.ac.uk
+ * @since 5.1.0
+ */
+public class UsedFilesRequestI extends UsedFilesRequest implements IRequest {
+
+    private static final long serialVersionUID = -1572148877023558009L;
+
+    private static final ImmutableMap<String, String> ALL_GROUPS_CONTEXT = ImmutableMap.of(Login.OMERO_GROUP, "-1");
+
+    private final PixelsService pixelsService;
+
+    private final Map<String, File> repositoryRoots = new HashMap<String, File>();
+
+    private Helper helper;
+    private Session session;
+    private Long filesetId;
+    private Long pixelsId;
+    private IFormatReader reader;
+
+    /**
+     * Construct a new used files request.
+     * @param pixelsService the pixels service
+     */
+    public UsedFilesRequestI(PixelsService pixelsService) {
+        this.pixelsService = pixelsService;
+    }
+
+    /* FIT USED FILES INTO A SINGLE REQUEST STEP */
+
+    @Override
+    public Map<String, String> getCallContext() {
+        return new HashMap<String, String>(ALL_GROUPS_CONTEXT);
+    }
+
+    @Override
+    public void init(Helper helper) {
+        this.helper = helper;
+        this.session = helper.getSession();
+        helper.setSteps(1);
```

Comment:
This is probably my only real concern. Mid to long-term we should either move away from single steps cmds, or separate the cmds into synchronous versions, since a 1-step cmd is a lot of overhead for no gain since the user can't cancel.